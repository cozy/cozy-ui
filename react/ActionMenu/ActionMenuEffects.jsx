import { useTheme } from '@material-ui/core'

import { getCssVariableValue } from '../utils/color'
import { useSetFlagshipUI } from '../hooks/useSetFlagshipUi/useSetFlagshipUI'

export const ActionMenuEffects = () => {
  const theme = useTheme()

  useSetFlagshipUI(
    {
      bottomBackground: theme.palette.background.paper,
      bottomTheme: 'dark',
      topOverlay: getCssVariableValue('overlay'), // TODO: refactor to semantic variable
      topBackground: theme.palette.background.paper,
      topTheme: 'light'
    },
    {
      bottomBackground:
        (document.getElementById('sidebar') &&
          getComputedStyle(document.getElementById('sidebar')).getPropertyValue(
            'background-color'
          )) ??
        theme.palette.background.paper,
      bottomTheme: 'dark',
      topOverlay: 'transparent',
      topBackground: theme.palette.background.paper,
      topTheme: 'dark'
    },
    'cozy-ui/ActionMenu'
  )

  return null
}
