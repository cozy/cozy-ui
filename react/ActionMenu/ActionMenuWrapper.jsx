import React from 'react'

import useBreakpoints from '../hooks/useBreakpoints'
import BottomDrawer from '../BottomDrawer'

import NotInlineWrapper from './NotInlineWrapper'

const ActionMenuWrapper = ({
  onClose,
  anchorElRef,
  popperOptions,
  placement,
  preventOverflow,
  children
}) => {
  const { isMobile } = useBreakpoints()

  if (isMobile) {
    return <BottomDrawer onClose={onClose}>{children}</BottomDrawer>
  }

  return (
    <NotInlineWrapper
      anchorElRef={anchorElRef}
      popperOptions={popperOptions}
      onClose={onClose}
      placement={placement}
      preventOverflow={preventOverflow}
    >
      {children}
    </NotInlineWrapper>
  )
}

export default ActionMenuWrapper
