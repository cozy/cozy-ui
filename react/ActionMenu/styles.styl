@require '../../stylus/components/action-menu'

.c-actionmenu
    @extends $actionmenu

.c-actionmenu--inline
    @extends $actionmenu--inline

.c-actionmenu-header
    @extends $actionmenu-header

.c-actionmenu-item
    @extends $actionmenu-item

.c-actionmenu-radio
    @extends $actionmenu-radio

