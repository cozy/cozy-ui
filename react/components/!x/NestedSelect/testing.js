import { ItemRow } from './NestedSelect'

export const findOptions = root => root.find(ItemRow)
