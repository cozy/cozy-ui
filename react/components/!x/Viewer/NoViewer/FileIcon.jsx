import React from 'react'

import Icon from '../../Icon'
import FileTypeBinIcon from '../../Icons/FileTypeBin'
import FileTypeCodeIcon from '../../Icons/FileTypeCode'
import FileTypeSheetIcon from '../../Icons/FileTypeSheet'
import FileTypeSlideIcon from '../../Icons/FileTypeSlide'
import FileTypeTextIcon from '../../Icons/FileTypeText'
import FileTypeZipIcon from '../../Icons/FileTypeZip'
import FileTypePdfIcon from '../../Icons/FileTypePdf'
import FileTypeFilesIcon from '../../Icons/FileTypeFiles'

const FileIcon = ({ type }) => {
  let icon

  switch (type) {
    case 'bin':
      icon = FileTypeBinIcon
      break
    case 'code':
      icon = FileTypeCodeIcon
      break
    case 'spreadsheet':
      icon = FileTypeSheetIcon
      break
    case 'slide':
      icon = FileTypeSlideIcon
      break
    case 'text':
      icon = FileTypeTextIcon
      break
    case 'zip':
      icon = FileTypeZipIcon
      break
    case 'pdf':
      icon = FileTypePdfIcon
      break
    default:
      icon = FileTypeFilesIcon
      break
  }

  return <Icon icon={icon} width={160} height={140} />
}

export default FileIcon
