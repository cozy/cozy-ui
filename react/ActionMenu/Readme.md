### Classic

```
import ActionMenu, { ActionMenuItem } from 'cozy-ui/transpiled/react/ActionMenu';
import Icon from 'cozy-ui/transpiled/react/Icon';

const showMenu = () => setState({ menuDisplayed: true })
const hideMenu = () => setState({ menuDisplayed: false });

<div>
  <button onClick={showMenu}>Show action menu</button>
  {state.menuDisplayed &&
    <ActionMenu
      onClose={hideMenu}>
      <ActionMenuItem left={<Icon icon='file' />}>Item 1</ActionMenuItem>
      <ActionMenuItem left={<Icon icon='right' />}>Item 2</ActionMenuItem>
      <ActionMenuItem left={<Icon icon='file' />}>Item 3</ActionMenuItem>
  </ActionMenu>}
</div>
```

### With Header

```
import ActionMenu, { ActionMenuItem, ActionMenuHeader } from './index';
import Icon from '../Icon';
import Filename from '../Filename';

const showMenu = () => setState({ menuDisplayed: true })
const hideMenu = () => setState({ menuDisplayed: false });

<div>
  <button onClick={showMenu}>Show action menu</button>
  {state.menuDisplayed &&
    <ActionMenu
      onClose={hideMenu}>
      <ActionMenuHeader>
        <Filename icon="file" filename="my_awesome_paper" extension=".pdf" />
      </ActionMenuHeader>
      <ActionMenuItem left={<Icon icon='file' />}>Item 1</ActionMenuItem>
      <ActionMenuItem left={<Icon icon='right' />}>Item 2</ActionMenuItem>
      <ActionMenuItem left={<Icon icon='file' />}>Item 3</ActionMenuItem>
  </ActionMenu>}
</div>
```

### With Header & onClick

```
import ActionMenu, { ActionMenuItem, ActionMenuHeader } from './index';
import Icon from '../Icon';
import Filename from '../Filename';

const showMenu = () => setState({ menuDisplayed: true })
const hideMenu = () => setState({ menuDisplayed: false });

<div>
  <button onClick={showMenu}>Show action menu</button>
  {state.menuDisplayed &&
    <ActionMenu
      onClose={hideMenu}>
      <ActionMenuHeader>
        <Filename icon="file" filename="my_awesome_paper" extension=".pdf" />
      </ActionMenuHeader>
      <ActionMenuItem onClick={() => alert('click')}left={<Icon icon='file' />}>Item 1</ActionMenuItem>
  </ActionMenu>}
</div>
```
