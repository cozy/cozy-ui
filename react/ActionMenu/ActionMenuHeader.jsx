import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'

import styles from './styles.styl'

export const ActionMenuHeader = ({ children, className }) => {
  return (
    <div className={cx(styles['c-actionmenu-header'], className)}>
      {children}
    </div>
  )
}

ActionMenuHeader.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string
}
