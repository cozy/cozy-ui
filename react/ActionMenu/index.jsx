import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Hammer from 'hammerjs'
import styles from './styles.styl'
import Overlay from '../Overlay'
import once from 'lodash/once'
import { Media, Bd, Img } from '../Media'

const TRANSITION_DURATION = 100 // need to be kept in sync with css

/**
 * Use an ActionMenu to show the user possible actions to the user
 * at the bottom of the screen in a modal way.
 *
 * - Animates in/out
 * - Possible to use Escape to close (thanks to `<Overlay />`)
 * - Possible to click outside to close (thanks to `<Overlay />`)
 * - Reacts to gestures with HammerJS
 */
class ActionMenu extends Component {
  constructor(props) {
    super(props)
    this.state = { closing: false }
  }

  componentDidMount() {
    this.initialAppear()
    this.attachEvents()
  }

  componentWillUnmount() {
    this.gesturesHandler.destroy()
    this.menuNode.style = '' // Drops the node style in case it gets recycled, see https://github.com/cozy/cozy-ui/pull/602
  }

  initialAppear() {
    this.turnTransitionsOff()
    this.applyTransformation(1)
    requestAnimationFrame(() => {
      this.turnTransitionsOn()
      this.applyTransformation(0)
    })
  }

  turnTransitionsOn() {
    this.menuNode.classList.add(styles['with-transition'])
  }

  turnTransitionsOff() {
    this.menuNode.classList.remove(styles['with-transition'])
  }

  attachEvents() {
    this.gesturesHandler = new Hammer.Manager(this.wrapperNode, {
      recognizers: [[Hammer.Pan, { direction: Hammer.DIRECTION_VERTICAL }]]
    })

    // to be completely accurate, `maximumGestureDelta` should be the difference between the top of the menu and the
    // bottom of the page; but using the height is much easier to compute and accurate enough.
    const maximumGestureDistance = this.menuNode.getBoundingClientRect().height
    // between 0 and 1, how far down the gesture must be to be considered complete upon release
    const minimumCloseDistance = 0.6
    // a gesture faster than this will dismiss the menu, regardless of distance traveled
    const minimumCloseVelocity = 0.6

    let currentGestureProgress = null

    this.gesturesHandler.on('panstart', () => {
      this.turnTransitionsOff()
      currentGestureProgress = 0
    })

    this.gesturesHandler.on('pan', e => {
      currentGestureProgress = e.deltaY / maximumGestureDistance
      this.applyTransformation(currentGestureProgress)
    })

    this.gesturesHandler.on('panend', e => {
      // Dismiss the menu if the swipe pan was bigger than the treshold,
      // or if it was a fast, downward gesture
      let shouldDismiss =
        e.deltaY / maximumGestureDistance >= minimumCloseDistance ||
        (e.deltaY > 0 && e.velocity >= minimumCloseVelocity)

      if (shouldDismiss) {
        if (currentGestureProgress >= 1) {
          // Menu is already hidden, close it right away
          this.close()
        } else {
          this.turnTransitionsOn()
          this.animateClose()
        }
      } else {
        this.turnTransitionsOn()
        this.applyTransformation(0)
      }
    })
  }

  /**
   * Applies a css trasnform to the element, based on the progress of the gesture
   * @param  {Float} progress - Amount of progress between 0 and 1
   */
  applyTransformation(progress) {
    // constrain between 0 and 1.1 (go a bit further than 1 to be hidden completely)
    progress = Math.min(1.1, Math.max(0, progress))
    this.menuNode.style.transform = 'translateY(' + progress * 100 + '%)'
  }

  animateClose = () => {
    this.setState({ closing: true })

    // we need to transition the menu to the bottom before dismissing it
    const close = once(() => {
      this.menuNode && this.menuNode.removeEventListener('transitionend', close)
      this.close()
    })

    this.menuNode &&
      this.menuNode.addEventListener('transitionend', close, false)
    // in case transitionend is not called, for example if the element is removed
    setTimeout(close, TRANSITION_DURATION)

    this.applyTransformation(1.1)
  }

  close = () => {
    this.setState({ closing: true })
    this.props.onClose()
  }

  handleMenuRef = menu => {
    this.menuNode = menu
  }

  handleWrapperRef = wrapper => {
    this.wrapperNode = wrapper
  }

  render() {
    const { children, className } = this.props
    const { closing } = this.state
    return (
      <div className={className} ref={this.handleWrapperRef}>
        <Overlay
          style={{ opacity: closing ? 0 : 1 }}
          onClick={this.animateClose}
          onEscape={this.animateClose}
        >
          <div className={styles['c-actionmenu']} ref={this.handleMenuRef}>
            {children}
          </div>
        </Overlay>
      </div>
    )
  }
}

ActionMenu.propTypes = {
  /** Extra class */
  className: PropTypes.string,
  /** What to do on close */
  onClose: PropTypes.func
}

const ActionMenuHeader = ({ children }) => {
  return <div className={styles['c-actionmenu-header']}>{children}</div>
}

const ActionMenuItem = ({ left, children, right }) => {
  return (
    <Media className={styles['c-actionmenu-item']}>
      {left && <Img className="u-mh-1">{left}</Img>}
      <Bd className={left ? 'u-mr-1' : 'u-mh-1'}>{children}</Bd>
      {right && <Img className="u-mr-1">{right}</Img>}
    </Media>
  )
}

export default ActionMenu
export { ActionMenuHeader, ActionMenuItem }
