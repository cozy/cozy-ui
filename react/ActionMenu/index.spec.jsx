import React from 'react'
import { mount } from 'enzyme'
import { render } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

import { fixPopperTesting } from '../Popper/testing'
import { BreakpointsProvider } from '../hooks/useBreakpoints'
import Typography from '../Typography'
import Icon from '../Icon'
import FileIcon from '../Icons/File'
import WarningIcon from '../Icons/Warning'

import ActionMenu, { ActionMenuItem, ActionMenuRadio } from './'

describe('ActionMenu', () => {
  fixPopperTesting()

  // The update-not-wrapping-in-act warning is disabled for ActionMenuWrapper since
  // we have not found how to remove the "update" happening when mounting the
  // ActionMenu.
  let originalConsoleError = console.error

  beforeEach(() => {
    // eslint-disable-next-line no-console
    console.error = function(msg, arg) {
      if (
        msg.includes('An update to %s inside a test was not wrapped in act') &&
        arg == 'ActionMenuWrapper'
      ) {
        return
      } else {
        return originalConsoleError.apply(this, arguments)
      }
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error = originalConsoleError
  })

  it('should support null children', async () => {
    const comp = mount(
      <BreakpointsProvider>
        <ActionMenu onClose={jest.fn()}>
          <ActionMenuItem>Item 1</ActionMenuItem>
          {null}
        </ActionMenu>
      </BreakpointsProvider>
    )
    // Remove an update was not wrapped in act() warning
    await act(async () => {})
    expect(
      comp
        .find(ActionMenuItem)
        .parent()
        .getElement()
    ).toMatchSnapshot()
  })

  it('should support auto-closing the menu', async () => {
    const closeMenu = jest.fn()
    const menuAction1 = jest.fn()
    const menuAction2 = jest.fn()
    const menuActionStoppingPropagation = e => {
      e.stopPropagation()
      menuAction2()
    }

    const comp = mount(
      <BreakpointsProvider>
        <ActionMenu onClose={closeMenu} autoclose>
          <ActionMenuItem onClick={menuAction1}>Item 1</ActionMenuItem>
          <ActionMenuItem onClick={menuActionStoppingPropagation}>
            Item 2
          </ActionMenuItem>
        </ActionMenu>
      </BreakpointsProvider>
    )

    await act(async () => {
      comp
        .find(ActionMenuItem)
        .at(1)
        .simulate('click')
    })
    expect(menuAction2).toHaveBeenCalled()
    expect(closeMenu).not.toHaveBeenCalled()

    act(() => {
      comp
        .find(ActionMenuItem)
        .at(0)
        .simulate('click')
    })
    expect(menuAction1).toHaveBeenCalled()
    expect(closeMenu).toHaveBeenCalled()
  })

  it('should render as expected', async () => {
    const { container } = render(
      <BreakpointsProvider>
        <ActionMenuItem
          left={<Icon icon={FileIcon} />}
          right={<Icon icon={WarningIcon} color="var(--errorColor)" />}
          onClick={() => alert('click')}
        >
          Item 1 with onclick action
        </ActionMenuItem>
        <ActionMenuItem left={<Icon icon={FileIcon} />}>
          Item 2 with dialog action
        </ActionMenuItem>
        <ActionMenuItem left={<ActionMenuRadio />}>Item 3</ActionMenuItem>
        <ActionMenuItem left={<Icon icon={FileIcon} />}>
          <Typography variant="body1" gutterBottom>
            Item 4
          </Typography>
          <Typography variant="caption" color="textSecondary">
            Descriptive text to elaborate on what item 3 does.
          </Typography>
        </ActionMenuItem>
      </BreakpointsProvider>
    )

    expect(container).toMatchSnapshot()
  })
})
