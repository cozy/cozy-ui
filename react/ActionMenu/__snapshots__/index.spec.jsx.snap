// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionMenu should render as expected 1`] = `
<div>
  <div
    class="styles__media___cSJMp styles__media--top___K9w0I styles__c-actionmenu-item___gODqd"
  >
    <div
      class="styles__img___3SHpG u-mh-1"
    >
      <svg
        class="styles__icon___23x3R"
        height="16"
        viewBox="0 0 16 16"
        width="16"
      >
        <defs>
          <path
            d="M1 1c0-.552.446-1 .998-1H10v5h5v10.004a1 1 0 01-1.007.996H2.007A1.002 1.002 0 011 15V1zm10 3V0l4 4h-4z"
            id="file_svg__a"
          />
        </defs>
        <use
          fill-rule="evenodd"
          xlink:href="#file_svg__a"
        />
      </svg>
    </div>
    <div
      class="styles__bd___1Uv-F u-mr-1 u-stack-xs"
    >
      Item 1 with onclick action
    </div>
    <div
      class="styles__img___3SHpG u-mr-1"
    >
      <svg
        class="styles__icon___23x3R"
        height="16"
        style="fill: var(--errorColor);"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="M7.022 1.736c.543-.959 1.428-.95 1.966 0l6.532 11.529c.543.958.094 1.735-1.016 1.735h-13C.399 15-.051 14.215.487 13.265l6.535-11.53zM7.004 13c0-.552.443-1 1-1 .552 0 1 .444 1 1 0 .553-.444 1-1 1-.553 0-1-.444-1-1zm0-7.997a.999.999 0 011-1.003c.552 0 1 .438 1 1.003v4.994a.999.999 0 01-1 1.003c-.553 0-1-.438-1-1.003V5.003z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
  <div
    class="styles__media___cSJMp styles__media--top___K9w0I styles__c-actionmenu-item___gODqd"
  >
    <div
      class="styles__img___3SHpG u-mh-1"
    >
      <svg
        class="styles__icon___23x3R"
        height="16"
        viewBox="0 0 16 16"
        width="16"
      >
        <defs>
          <path
            d="M1 1c0-.552.446-1 .998-1H10v5h5v10.004a1 1 0 01-1.007.996H2.007A1.002 1.002 0 011 15V1zm10 3V0l4 4h-4z"
            id="file_svg__a"
          />
        </defs>
        <use
          fill-rule="evenodd"
          xlink:href="#file_svg__a"
        />
      </svg>
    </div>
    <div
      class="styles__bd___1Uv-F u-mr-1 u-stack-xs"
    >
      Item 2 with dialog action
    </div>
  </div>
  <div
    class="styles__media___cSJMp styles__media--top___K9w0I styles__c-actionmenu-item___gODqd"
  >
    <div
      class="styles__img___3SHpG u-mh-1"
    >
      <label
        class="styles__c-input-radio___XJQt1 styles__c-actionmenu-radio___Km9uj"
      >
        <input
          type="radio"
          value=""
        />
        <span />
      </label>
    </div>
    <div
      class="styles__bd___1Uv-F u-mr-1 u-stack-xs"
    >
      Item 3
    </div>
  </div>
  <div
    class="styles__media___cSJMp styles__media--top___K9w0I styles__c-actionmenu-item___gODqd"
  >
    <div
      class="styles__img___3SHpG u-mh-1"
    >
      <svg
        class="styles__icon___23x3R"
        height="16"
        viewBox="0 0 16 16"
        width="16"
      >
        <defs>
          <path
            d="M1 1c0-.552.446-1 .998-1H10v5h5v10.004a1 1 0 01-1.007.996H2.007A1.002 1.002 0 011 15V1zm10 3V0l4 4h-4z"
            id="file_svg__a"
          />
        </defs>
        <use
          fill-rule="evenodd"
          xlink:href="#file_svg__a"
        />
      </svg>
    </div>
    <div
      class="styles__bd___1Uv-F u-mr-1 u-stack-xs"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        Item 4
      </p>
      <span
        class="MuiTypography-root MuiTypography-caption MuiTypography-colorTextSecondary"
      >
        Descriptive text to elaborate on what item 3 does.
      </span>
    </div>
  </div>
</div>
`;

exports[`ActionMenu should support null children 1`] = `
<div
  className="styles__c-actionmenu___22Fp1 styles__c-actionmenu--inline___1SXZa"
  onClick={[Function]}
  onTouchEnd={[Function]}
>
  <ActionMenuItem
    contentSpacing="xs"
  >
    Item 1
  </ActionMenuItem>
</div>
`;
